### PR TITLE
python 2.7.14

### PR DESCRIPTION
**Pull request by chuanconggao**
- [X] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [X] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/python.rb [hunk 1]
```diff
@@ -1,8 +1,8 @@
 class Python < Formula
   desc "Interpreted, interactive, object-oriented programming language"
   homepage "https://www.python.org/"
-  url "https://www.python.org/ftp/python/2.7.13/Python-2.7.13.tar.xz"
-  sha256 "35d543986882f78261f97787fd3e06274bfa6df29fac9b4a94f73930ff98f731"
+  url "https://www.python.org/ftp/python/2.7.14/Python-2.7.14.tar.xz"
+  sha256 "71ffb26e09e78650e424929b2b457b9c912ac216576e6bd9e7d204ed03296a66"
   revision 1
```
[the remark this answers is not in the document]
Comment (chuanconggao) [reply Formula/python.rb]: Thanks. I missed that line.

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by apjanke** [comment]
`sip`, `vim`, and `macvim` are failing their tests with this.

For `sip`:

```
        Testing sip
/usr/bin/sandbox-exec -f /tmp/homebrew20170918-60079-qlyb39.sb /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/sip.rb --verbose
==> /usr/bin/clang++ -shared -Wl,-install_name,/tmp/sip-test-20170918-60080-1dot5rz/libtest.dylib -o libtest.dylib test.cpp
==> /usr/local/Cellar/sip/4.19.3_1/bin/sip -b test.build -c . test.sip
==> python generate.py
==> make -j1 clean all
rm -f test.so
rm -f siptestcmodule.o
rm -f siptestTest.o
c++ -c -pipe -fPIC -Os -Wall -W -DNDEBUG -I. -I/usr/local/Cellar/sip/4.19.3_1/include -I/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7 -o siptestcmodule.o siptestcmodule.cpp
In file included from siptestcmodule.cpp:7:
In file included from ./sipAPItest.h:10:
/usr/local/Cellar/sip/4.19.3_1/include/sip.h:32:10: fatal error: 'Python.h' file not found
#include <Python.h>
         ^~~~~~~~~~
1 error generated.
make: *** [siptestcmodule.o] Error 1
Error: sip: failed
Failed executing: make -j1 clean all
```

I can reproduce this `sip` test failure locally by installing `python` from this PR. Rebuilding `sip` from source fixed it. Might just need a revision bump here.

Same with `vim`. I get a repro of this:
```
$ brew test vim
Testing vim
==> /usr/local/Cellar/vim/8.0.1100_1/bin/vim -T dumb -s commands.vim test.txt
Error: vim: failed
<"hello world"> expected but was
<"">.
```
when installing from bottle. A rebuild fixes it. Doing a `grep -r 2.7.13 *` in the `vim` keg gets several matches, including in the binaries, so I'm guessing the exact `python` version is baked in there.

The `macvim` test is failing because the test itself is baking in a check for the exact path to the Cellar location of the `python` that MacVim was built against.

```
  test do
    ENV.prepend_path "PATH", Formula["python"].opt_libexec/"bin"
    # Simple test to check if MacVim was linked to Python version in $PATH
    if build.with? "python"
      system_framework_path = `python-config --exec-prefix`.chomp
      assert_match system_framework_path, `mvim --version`
    end
  end
```

Because of this:
```
$ /usr/local/opt/python/libexec/bin/python-config --exec-prefix
/usr/local/Cellar/python/2.7.14/Frameworks/Python.framework/Versions/2.7
```
it ends up searching for the exact patch version string in the output of `mvim --version`, even though the linking in MacVim uses `/usr/local/opt/python/.../2.7/...` only down to the minor version. Don't know if that will cause actual breakage.

Regardless, a rebuild of MacVim after installing the new `python` fixes it.

Sounds like revision bumps all around should take care of this.

**Comment by ilovezfs** [comment]
@apjanke what does your local `brew config` show?

**Comment by ilovezfs** [comment]
`sip` failure is unrelated to this PR: https://jenkins.brew.sh/job/Homebrew%20Testing/178/version=sierra/console

I think it's caused by the Xcode 8 -> Xcode 9 upgrade, since `sip` has hard coded the path

```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7
```

but in Xcode 9, this should be

```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7
```

We could try to reduce the number of revision bumps by using the versioned path

```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7
```
but we could do that at any time.

Similarly it's hard coding
```
/usr/local/Cellar/python3/3.6.2
```
which we could replace with `/usr/local/opt/python3`.

**Comment by ilovezfs** [comment]
Also, instead of
```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7
```
we could probably even use `/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7`.

**Comment by DomT4** [comment]
Might be good to update `setuptools` and `wheel` here as well.

**Comment by cclauss** [comment]
@chuanconggao  Are you able to update `setuptools` to 36.5.0 and `wheel` to 0.30.0?  This is a popular formula so it would be good to get this PR accepted.

**Comment by chuanconggao** [comment]
@cclauss Sorry for the delay. I have updated `setuptools` and `wheel` to their latest versions.

**Comment by DomT4** [comment]
Going to need `vim` & `macvim` revision bumps here as well. And whatever ILZ wants done with `sip`. After that we should be good, I think?

**Comment by chuanconggao** [comment]
@DomT4 How do I do revision bumps for `vim` and `macvim`? Just add `revision 1` to their respective formulas?

**Comment by DomT4** [comment]
```diff
diff --git a/Formula/macvim.rb b/Formula/macvim.rb
index adda7fe2c3..add3ded6f5 100644
--- a/Formula/macvim.rb
+++ b/Formula/macvim.rb
@@ -5,6 +5,8 @@ class Macvim < Formula
   url "https://github.com/macvim-dev/macvim/archive/snapshot-137.tar.gz"
   version "8.0-137"
   sha256 "2f2b20c80f887c5f26dad42540f047e43928f8c1cb217b6874c8d5602eceb620"
+  revision 1
+
   head "https://github.com/macvim-dev/macvim.git"
 
   bottle do
diff --git a/Formula/vim.rb b/Formula/vim.rb
index 368d24f475..a8aa93ee5f 100644
--- a/Formula/vim.rb
+++ b/Formula/vim.rb
@@ -3,7 +3,7 @@ class Vim < Formula
   homepage "https://vim.sourceforge.io/"
   url "https://github.com/vim/vim/archive/v8.0.1100.tar.gz"
   sha256 "4737599b2153388b1d8621a84f2ef2162474e17c7179375b60e17f551902d8c1"
-  revision 1
+  revision 2
 
   head "https://github.com/vim/vim.git"
```

Essentially. Although if `vim` has a new version available feel free to version bump it instead I guess, that'd work too.

**Comment by chuanconggao** [comment]
@DomT4 Thanks. I just did the two revision bumps.

**Comment by apjanke** [comment]
> Also, instead of
> ```
> /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7
> ```
> we could probably even use
> ```
> /System/Library/Frameworks/Python.framework/Versions/2.7/include/python2.7
> ```

Yeah, could and should imho, because some people won't even have `/Applications/Xcode.app` installed. Even devs: e.g. mine are all versioned like `/Applications/Xcode-8.3.3.app`. `brew test sip` throws the same failure for me even without this PR.

I took a look at `sip`'s `configure.py`, and there doesn't seem to be a simple command line option to point at a different Python header directory. But I'm no Python expert.

I know you hate red tests, @ilovezfs, but I think we could merge this one as is.

**Comment by ilovezfs** [comment]
Yeah :-1: to merging while red. Basically there's no reason to believe it's not going to be red next time unless this is fixed now.

**Comment by apjanke** [comment]
I'm looking at `sip`'s installation stuff now. Going to try getting it to work with a custom configuration file as though it were a cross-compilation.

**Comment by apjanke** [comment]
Fix for `sip`: https://github.com/Homebrew/homebrew-core/pull/18490

Feedback welcome, since this is kind of a hack and I'm no Python expert.

**Comment by ilovezfs** [comment]
:green_apple:

**Comment by ilovezfs** [comment]
Thanks @chuanconggao! Shipped! Just for future reference it should be one formula per commit.

**Pull request closed without merging**